### ISSUE TITLE
Question: Graphical requirements of the library

### ISSUE THREAD

**Issue by dave-br** [Radirius/Mercury#24]
I'm hitting the two exceptions below when running a very simple starter project via Eclipse.  This same project works great on a couple personal computers I tried, but I'm hitting these exceptions when trying to run the project for one of my students in a high school computer lab.  At first, I thought it might be due to some of the locked-down policies on these school computers, but I still hit these exceptions when logged in as myself (and I'm a local admin on the computer).

A colleague suggested the issue might be that Mercury / lwjgl is expecting a minimum OpenGL version that the computer doesn't support.  I did find some lwjgl code here, which I can try running next week, assuming this properly reports what OpenGL version is supported on the computer: http://wiki.lwjgl.org/wiki/Version_selection

Could you confirm what the OpenGL requirements are, and a good way to validate?  Or can you come up with other reasons these exceptions might be encountered?

Below is the output reported by Eclipse, followed by the source code for the single class in the project.  (Again, this code works GREAT on the two non-school PCs I tried.)

Thanks for your help!

Output:

```
Mercury Game Library (In-Dev)
Website: http://mercurylib.com/
-------------------------------

You're running a non-stable build of Mercury!
If you run into any issues, please leave an issue on GitHub or make a end on the forum. 
org.lwjgl.LWJGLException: Could not create context (WGL_ARB_create_context)
	at org.lwjgl.opengl.WindowsContextImplementation.nCreate(Native Method)
	at org.lwjgl.opengl.WindowsContextImplementation.create(WindowsContextImplementation.java:50)
	at org.lwjgl.opengl.ContextGL.<init>(ContextGL.java:132)
	at org.lwjgl.opengl.Display.create(Display.java:850)
	at org.lwjgl.opengl.Display.create(Display.java:797)
	at com.radirius.mercury.framework.Window.initDisplay(Window.java:161)
	at com.radirius.mercury.framework.Core.run(Core.java:175)
	at com.radirius.mercury.framework.Core.start(Core.java:48)
	at MyFirstGame.main(MyFirstGame.java:16)
Exception in thread "main" java.lang.RuntimeException: No OpenGL context found in the current thread.
	at org.lwjgl.opengl.GLContext.getCapabilities(GLContext.java:124)
	at org.lwjgl.opengl.GL11.glEnable(GL11.java:1012)
	at com.radirius.mercury.graphics.Graphics.initGraphics(Graphics.java:978)
	at com.radirius.mercury.framework.Core.run(Core.java:177)
	at com.radirius.mercury.framework.Core.start(Core.java:48)
	at MyFirstGame.main(MyFirstGame.java:16)
```

Source code:

```java
import com.radirius.mercury.framework.*;
import com.radirius.mercury.graphics.*;
import com.radirius.mercury.math.geometry.Line;

public class MyFirstGame extends Core {

    public MyFirstGame(CoreSetup coreSetup) {
        super(coreSetup);
    }

    public static void main(String[] args) {
        CoreSetup coreSetup = new CoreSetup("My First Game");
        coreSetup.width = 800;
        coreSetup.height = 600;

        new MyFirstGame(coreSetup).start();
    }

    public void init() {
    
    }

    public void update() {
    
    }

    public void render(Graphics g) 
    {
        for (int i=0; i < 800; i += 6)
        {
            Line line = new Line(400, 300, i, 0);
            g.drawLine(line);
        }
    
    }

    public void cleanup() {
    
    }
}
```

**Comment by sriharshachilakapati** [comment]
This library requires OpenGL 3.3 in core profile on Windows and Linux, and 3.2 on Mac. We use VAOs from (3.0) and in shaders we use the GLSL layout qualifier (GLSL 330). In theory, any decent GPU that is manufactured after 2008 should support these requirements, it even works on laptops. It maybe that your school systems are older, maybe from before 2008 (that's almost 8 years).

**Comment by dave-br** [comment]
Thanks for the info!  I'll investigate further what exactly is running on the school computers.

**Comment by dave-br** [comment]
Follow up...

The computers do appear to be a bit old.  They're apparently using the on-board Intel HD Graphics card, with Intel Core i5 CPU 650 @ 3.2GHz (family 6, model 37).  This apparently makes them "1st generation" Core i5's, which means, according to http://www.intel.com/support/graphics/sb/CS-033757.htm, that the corresponding Intel HD Graphics only support up to opengl 2.1.

So... is there any way to configure Mercury or lwjgl to give me a subset of capabilities and work nicely with 2.1?  For our purposes, we're just doing some light 2D drawing, with possibly some explosion effects, but not much animation.  (It's just a Battleship game, where items are mostly static and not running around.)

I did find a way that lwjgl would attempt to gracefully fall back to software acceleration, by running with VM parameter -Dorg.lwjgl.opengl.Display.allowSoftwareOpenGL=true, but that did not help.  I tried & verified at run-time that the flag caused the property to be set, but I hit the exact same exception.

Thanks for your help!

**Comment by sriharshachilakapati** [comment]
Unfortunately, no. With just OpenGL, we can't roll it back to 2.1, as some features we are using nowadays are missing in old version, which is deprecated. And if we had to roll it back, it's a lot of work to do, with no real value to it, as the Steam hardware survey says that 97% of people have access to OpenGL 3.3 features.

What you tried (running with `-Dorg.lwjgl.opengl.Display.allowSoftwareOpenGL=true`) will only work on compatibility profile, allowing the context provider to provide software OpenGL implementations such as `Mesa3D` which is common on some linux distributions. Core profiles must come from real drivers.

Indeed, Mercury started at version 1.1, but we soon removed all the deprecated parts of it. I think I have answered your question, we are not going back, sorry, as that is a heavy job and also we don't want any deprecated technologies.
